arbos/programs/cache.go

```diff
@@ -0,0 +1,32 @@
+// Copyright 2024, Offchain Labs, Inc.
+// For license information, see https://github.com/OffchainLabs/nitro/blob/master/LICENSE
+
+package programs
+
+import "github.com/ethereum/go-ethereum/common"
+
+type RecentPrograms struct {
+	queue []common.Hash
+	items map[common.Hash]struct{}
+}
+
+func NewRecentProgramsTracker() *RecentPrograms {
+	return &RecentPrograms{
+		queue: make([]common.Hash, 0, initialTxCacheSize),
+		items: make(map[common.Hash]struct{}, initialTxCacheSize),
+	}
+}
+
+func (p *RecentPrograms) Insert(item common.Hash, params *StylusParams) bool {
+	if _, ok := p.items[item]; ok {
+		return true
+	}
+	p.queue = append(p.queue, item)
+	p.items[item] = struct{}{}
+
+	if len(p.queue) > int(params.TxCacheSize) {
```

Comment:
If access pattern is (A,B,A,C,A,D,A,E....) first to be evicted will be A.
1) since this is inside a transaction - can't we have a reasonable upper-limit that means we don't really need to worry about eviction at all? (+possibly put some hardcoded upper limit that can't be reached)
2) since this is inside memory - if we do care about eviction we can probably have something a little more complicated that will be more LRU

Reply:
Sure we can do the sequence number idea in memory. I'm about to push a commit to make the scoping per-block too, and then have Rust maintain a super-set :)

Reply:
Switched to an LRU cache in Geth :)